hbase-common/src/main/java/org/apache/hadoop/hbase/io/util/BlockIOUtils.java

```diff
@@ -284,6 +311,10 @@ private static boolean preadWithExtraDirectly(ByteBuff buff, FSDataInputStream d
         throw e;
       }
       if (ret < 0) {
+        if (remain <= extraLen) {
+          // break for the "extra data" when hitting end of stream and remaining is necessary
+          break;
```

Comment:
you got it right. e.g. if there is a case that we read a HFile does not have the trailer, and just end with the last data block, we're hitting here. 

I added that because I found this logic exists already as part of BlockIOUtils#readWithExtraOnHeap (see below).

https://github.com/apache/hbase/blob/f6e9d3e1dd6546663319c436286ab58dd0e731ed/hbase-common/src/main/java/org/apache/hadoop/hbase/io/util/BlockIOUtils.java#L154-L156 

I thought this is good for handling some case that we don't expect automatically, but we can remove it to let the premature EOF throws back to the client.

Reply:
That's a fair point -- I would have expected that we were very explicit and did exactly what the caller said. I think you did the right thing. May be good to capture that in a comment :)